blog/2023-04-18-magazine-003.md

```diff
@@ -0,0 +1,325 @@
+
+That's why Wing has a built-in testing mechanism, that we've expanded to support testing your app in the cloud as well.
+Let's first recap how unit tests work in Wing. Here's a demo app I've written that manages a list of tasks using a bucket and counter:
+
+```js
+// task_list.w
+bring cloud;
+
+resource TaskList {
+  data: cloud.Bucket;
+  counter: cloud.Counter;
+  init() {
+    this.data = new cloud.Bucket();
+    this.counter = new cloud.Counter();
+  }
+
+  inflight add_task(title: str): str {
+    let id = "task-${this.counter.inc()}";
+    this.data.put(id, title);
+    return id;
+  }
+
+  inflight get_task(id: str): str {
+    let title = this.data.get(id);
+    return title;
+  }
+}
+```
+
+To write tests for this app, we can simply create new `cloud.Function`'s whose **resource id's start with `test` or `test:`**
+
+```js
+// task_list.w (extended)
+let task_list = new TaskList();
+
+new cloud.Function(inflight () => {
+  let id1 = task_list.add_task("get avocados");
+  let id2 = task_list.add_task("walk my cat");
+  assert(id1 != id2);
+}) as "test:add multiple tasks";
+
+new cloud.Function(inflight () => {
+  let id = task_list.add_task("buy a new hat");
+  let result = task_list.get_task(id);
+  assert(result == "buy a new hat");
+}) as "test:add and get task";
+```
+
+Thanks to recent work by our amazing console team, if we open the [Wing Console](https://docs.winglang.io/getting-started/installation#wing-console) with `wing it task_list.w`, we're now able to run the tests against the Wing simulator using the new testing panel:
+
+Before running             |  After running
+:-------------------------:|:-------------------------:
+![testing-panel](./assets/testing-panel.png)  |  ![testing-panel-success](./assets/testing-panel-success.png)
+
+> *By the way, did we mention the Wing Console now has a dark mode? 🌙*
+
+But we can also run the tests in the cloud by using the `--target` flag on the `wing test` CLI command. For example, to run the tests in AWS, we can run `wing test --target tf-aws task_list.w`:
+
+```bash
+$ wing test -t tf-aws hello.w
+✔ Compiling to tf-aws...
+✔ terraform init
+✔ terraform apply
+✔ Setting up test runner...
+✔ Running tests...
+pass ─ hello.tfaws » root/Default/env0/test:add multiple tasks
+pass ─ hello.tfaws » root/Default/env1/test:add and get task  
+✔ terraform destroy
+```
+
+How cool is that? Now you can test your app in the cloud as well as in the simulator, all from the comfort of your terminal.
+
+There is still a lot of work ahead of us - so far we only support cloud tests on AWS, so we would like to support more cloud providers. But we would like to make it easier to view logs from your tests, and to debug your app when tests fail. We're excited to see what you build with this new feature!
+
+### New SDK resources and methods
+
+Wing's standard library of resources is rapidly growing!
+Here are the latest additions:
+
+#### `cloud.Api`
```

Comment:
Lets mention the api resource co-creators Raphael and you here

Reply:
Done